### ISSUE TITLE
initial use: tictactoe example: Fatal: ''iostream'' file not found

### ISSUE THREAD

**Issue by PhilipOakley** [CoatiSoftware/Sourcetrail#790]
Hi, This is an initial usage report.. I'd seen the news on [CodeProject](https://www.codeproject.com/News/48256/Sourcetrail-is-now-free-and-open-source-software.aspx), read the blog and looked at the video, and thought 'This looks great. It will help me.'
(I'm on Windows 10 x64)

When I load the tictactoe example project I get a Fatal error that the include file "<iostream>" (line 3 of io.cpp) is not found.

It is not clear (from the messages/top level documentation) how to fix this fatal error (Fatal sounds bad...). However one can still navigate the code's 'source-trail' :wink:

On an earlier (first) attempt with the code for git (https://github.com/PhilipOakley/git), loaded via the custom filter (more below) I had had the same error for a similar #include file with a terse 'fix this' message, but nothing to say how to fix it.
```
        1 errors (1 fatal)

	Warning: Your project has fatal errors, which cause
	  a lot of missing information in affected files.
	  Try to resolve them!
```

 I'd even added the referenced include file to the project and done a full refresh (shift+F5) and had the same error.

However, on doing a full refresh with tictactoe the problem is gone. But for my attempted Git project the error is still there.

1. The primary request for this issue is that the Fatal error message be improved so as to direct the user to a how to fix it documentation article.  (which may end up as a list of known 'how to fix' items, just to get the new starts over the initial issues)

2. The second request is that the Custom source group opens a window that is big enough to see all the boxes. I got confused (because of the awkward Git-for-Windows project set up) about selecting the project files as I had no IDE for the project, and it (git sdk) runs `make` on `gcc` in `bash` on`MSYS2\MINGW`, so too many incompatibility layers... So, I simply ended up with using the custom source group, and then using Explorer (selecting *.c, the *.h) to drag and drop the files into the group.

3. Hmm, so this also shows that the intro videos need something to cover when/if the *.h files should be added, and what/why/how source groups affect the processing (It maybe that you just need to say that for simple initial projects, everything goes in the one group and that later one can see how they work). Mainly , there should be an easy way (video?) for including a simple project (of a few files, inc *.h) that's not in an IDE.

Summary, it's still looking good. 

various "Empty C(pp) Source Group" entry methods (when starting a new project at least) include not only an "All files

**Comment by egraether** [comment]
Thanks for the feedback!

We have a video on how to fix errors during project setup, which is also linked in the documentation: https://www.youtube.com/watch?v=7fguPwKR_7Y&t=6s

We are aware that many users have issues during project setup. It is often not understood that we did not write the error messages ourselves. We use the LLVM/Clang compiler for indexing and **all error messages are standard error messages of the Clang compiler**. You would get the same errors when compiling the source code with the options you supplied in your Sourcetrail project with Clang.

We show a dialog after the first indexing run that yields errors, which explains this. It is also accessible at 2 locations in the UI:

<img width="711" alt="Screenshot 2019-11-25 at 22 17 20" src="https://user-images.githubusercontent.com/201867/69580042-09caf200-0fd4-11ea-8574-9f22c9172d19.png">
<img width="1280" alt="Screenshot 2019-11-25 at 22 15 11" src="https://user-images.githubusercontent.com/201867/69580048-0fc0d300-0fd4-11ea-9086-d86875135fa2.png">


I opened issues #791 #792 based on your feedback. #727 could have also made things easier for you. I will close this one though. Thanks for your feedback!

**Comment by PhilipOakley** [comment]
> We have a video on how to fix errors during project setup, which is also linked in the documentation: https://www.youtube.com/watch?v=7fguPwKR_7Y&t=6s

Thanks, that video helped. 

I did remember that Git-for-Windows does have a pre-prepared .sln project at https://github.com/git-for-windows/git/tree/vs/master so I checked that out into my working directory, acquired the [Sourcetrail VS plug-in](https://marketplace.visualstudio.com/items?itemName=vs-publisher-1208751.SourcetrailExtension) (which installed very nicely, thank you), and created the Compilation Database (all good so far).

However the one unexpected item (1352 errors!) was the need for "Preprocessor Definitions" to be set, which I didn't see mentioned in the video, and don't appear to have a Source Group entry for.

The 54 (!) Git preprocessor definitions are (from VS properties, with `;` changed to crlf)
```
WIN32
_DEBUG
DEFAULT_GIT_TEMPLATE_DIR="share/git-core/templates"
DEFAULT_HELP_FORMAT="html"
DETECT_MSYS_TTY
GIT_HOST_CPU="x86_64"
GIT_HTML_PATH="share/doc/git-doc"
GIT_INFO_PATH="share/info"
GIT_MAN_PATH="share/man"
HAVE_ALLOCA_H
HAVE_STRING_H
HAVE_WPGMPTR
MMAP_PREVENTS_DELETE
NATIVE_CRLF
NDEBUG
NOGDI
NO_GETTEXT
NO_INTTYPES_H
NO_IPV6
NO_LIBGEN_H
NO_MEMMEM
NO_MKDTEMP
NO_POLL
NO_POLL_H
NO_POSIX_GOODIES
NO_PREAD
NO_SETENV
NO_STRCASESTR
NO_STRLCPY
NO_STRTOUMAX
NO_ST_BLOCKS_IN_STRUCT_STAT
NO_SYMLINK_HEAD
NO_SYS_POLL_H
NO_UNIX_SOCKETS
OBJECT_CREATION_MODE=1
PAGER_ENV="LESS=FRX LV=-c"
PROTECT_NTFS_DEFAULT=1
RUNTIME_PREFIX
SHA1DC_CUSTOM_INCLUDE_SHA1_C="cache.h"
SHA1DC_CUSTOM_INCLUDE_UBC_CHECK_C="git-compat-util.h"
SHA1DC_INIT_SAFE_HASH_DEFAULT=0
SHA1DC_NO_STANDARD_INCLUDES
SHA1_DC
SHA256_BLK
SHELL_PATH="/bin/sh"
STRIP_EXTENSION=".exe"
UNRELIABLE_FSTAT
USE_CURL_FOR_IMAP_SEND
USE_WIN32_MMAP
WIN32
_CONSOLE
_CRT_NONSTDC_NO_DEPRECATE
_CRT_SECURE_NO_WARNINGS
__USE_MINGW_ACCESS
%(PreprocessorDefinitions)
```

How should pre-processor definitions be entered here? 
(e.g. additional compiler flags? but which flags, VS has lost/hides that aspect)

And could something (containing the words "preprocess definitions" for easy search;-) be added to the help pages to explain the solution

**Comment by PhilipOakley** [comment]
Story continued. 

I determined that (one of) the [correct preprosessor prefix](https://docs.microsoft.com/en-us/cpp/build/reference/d-preprocessor-definitions?view=vs-2019) was "-D " (with conversion from VS to Clang prefix..)

I had to parse the full list of files (access from the overview) to identify which ones had not been fully parsed (little 'x' in the top left of their file icon), and then go to the Visual Studio properties of those file/projects (Git's got 19 projects in the solution) and copy the preprocessor directives.

The sets of preprocessor directives from VS need the `;` separator changed to line feed (I used Notepad++ which was to hand), and then filtered for uniqueness (cheated: filtered the list in Excel), pasted back to Notepad++ and then pre-pended the "-D" before finally pasting the list into the "Additional Compiler Flags" box (click on the writing pen icon) to add them.

Saved and re-indexed. Still some problems.

Went back to VS and actually built the solution (see earlier, the instructions just said to use the Sourcetrail >> Create Compilation Database menu... no mention of building:wink:),

And then re-created the compilation database, and re-index _again_. 
(the actual build had 4 link errors to a library, probably updated, that somehow I don't have.)

The re-indexing now resulted in just 4 errors (not fatal), but it look like it's a compiler standard issue (#679).

Pretty close now. Hopefully the story helps.
